shell/browser/hid/hid_chooser_context.cc

```diff
@@ -99,6 +105,66 @@ void HidChooserContext::GrantDevicePermission(
   }
 }
 
+void HidChooserContext::RevokeDevicePermission(
+    const url::Origin& origin,
+    const device::mojom::HidDeviceInfo& device,
+    content::RenderFrameHost* render_frame_host) {
+  DCHECK(base::Contains(devices_, device.guid));
+  if (CanStorePersistentEntry(device)) {
+    RevokePersistentDevicePermission(origin, device, render_frame_host);
+  } else {
+    RevokeEphemeralDevicePermission(origin, device);
+  }
+  auto* web_contents =
+      content::WebContents::FromRenderFrameHost(render_frame_host);
+
+  api::Session* session =
+      api::Session::FromBrowserContext(web_contents->GetBrowserContext());
+  if (session) {
+    v8::Isolate* isolate = JavascriptEnvironment::GetIsolate();
+    v8::HandleScope scope(isolate);
+    v8::Local<v8::Object> details = gin::DataObjectBuilder(isolate)
+                                        .Set("device", device.Clone())
+                                        .Set("frame", render_frame_host)
+                                        .Build();
+    session->Emit("hid-device-revoked", details);
+  }
+}
+
+void HidChooserContext::RevokePersistentDevicePermission(
+    const url::Origin& origin,
+    const device::mojom::HidDeviceInfo& device,
+    content::RenderFrameHost* render_frame_host) {
+  auto* web_contents =
+      content::WebContents::FromRenderFrameHost(render_frame_host);
+  auto* permission_helper =
+      WebContentsPermissionHelper::FromWebContents(web_contents);
+  permission_helper->RevokeHIDDevicePermission(
+      origin, DeviceInfoToValue(device), render_frame_host);
+  RevokeEphemeralDevicePermission(origin, device);
+}
+
+void HidChooserContext::RevokeEphemeralDevicePermission(
+    const url::Origin& origin,
+    const device::mojom::HidDeviceInfo& device) {
+  auto it = ephemeral_devices_.find(origin);
+  if (it != ephemeral_devices_.end()) {
+    std::set<std::string>& devices = it->second;
+    for (auto guid = devices.begin(); guid != devices.end();) {
+      DCHECK(base::Contains(devices_, *guid));
+
+      if (devices_[*guid]->physical_device_id != device.physical_device_id) {
```

Comment:
If devices_ doesn't have *guid this will correctly DCHECK on line 154; but if we're not in debug mode it will crash on 156 in a nullptr dereference

Reply:
@ckerr this is copied from upstream here: https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/hid/hid_chooser_context.cc;drc=d1142df38072bfc3128ff643cf982923b2b161c5;l=431.  I suppose we could change the DCHECK to a CHECK if that's better?

Reply:
Your call. It'll crash either way, so I guess on balance it's _slightly_ better to follow upstream?

_If_ this could actually be nullptr the fix would be to upstream a check